fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -179,69 +179,126 @@ private IndexingByIndex getIndexerByIndex() {
-                        LogMessageKeys.INDEX_NAME, common.getIndex().getName(),
-                        LogMessageKeys.INDEX_VERSION, common.getIndex().getLastModifiedVersion(),
-                        LogMessageKeys.INDEXER_ID, common.getUuid(),
-                        LogMessageKeys.INDEX_STATE_PRECONDITION, common.getIndexStatePrecondition(),
-                        LogMessageKeys.ACTUAL_TYPE, conflictingIndexingTypeStamp
-                ).toString());
+                        LogMessageKeys.CURR_ATTEMPT, attemptCount,
+                        LogMessageKeys.INDEX_STATE_PRECONDITION, indexStatePrecondition,
+                        LogMessageKeys.ACTUAL_TYPE, conflictingIndexingTypeStamp)
+                        .addKeysAndValues(common.indexLogMessageKeyValues())
+                        .toString());
             }
-            if (common.getIndexStatePrecondition() == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY) {
+
+            if (indexStatePrecondition == IndexStatePrecondition.BUILD_IF_DISABLED_CONTINUE_BUILD_IF_WRITE_ONLY) {
+                // Make best effort to finish indexing. Attempt continuation of the previous method
                 indexer = null;
-                // now, match the policy to the previous run
+                // Now, match the policy to the previous run
                 IndexBuildProto.IndexBuildIndexingStamp.Method method = conflictingIndexingTypeStamp.getMethod();
-                if ( method == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
+                if (method == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
+                    // Partly built by records. The fallback indicator should handle the policy
                     fallbackToRecordsScan = true;
-                    return fallbackFunction.get();
+                    return indexingLauncher(indexingFunc, attemptCount);
                 }
                 if (method == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_INDEX) {
+                    // Partly built by index. Retry with the old policy, but preserve the requested policy - in case the old one fails.
                     Object sourceIndexSubspaceKey = decodeSubspaceKey(conflictingIndexingTypeStamp.getSourceIndexSubspaceKey());
+                    IndexingPolicy origPolicy = indexingPolicy;
                     indexingPolicy = IndexingPolicy.newBuilder()
                             .setSourceIndexSubspaceKey(sourceIndexSubspaceKey)
-                            .forbidRecordScan() // no third chance
+                            .setForbidRecordScan(origPolicy.forbidRecordScan)
                             .build();
-                    return fallbackFunction.get();
+                    return indexingLauncher(indexingFunc, attemptCount, origPolicy);
                 }
-                // None of the above? let it fall through
+                // No other methods (yet). This line should never be reached.
```

Comment:
Added log error + recovery. 
(Not an exception because it's related to the existing type stamp (of a previous indexing or another issue) and will be only repeated when called again.)